### PR TITLE
First wave experimental ordering

### PR DESCRIPTION
**Pull request by Amoki**

### CODE REVIEW

File: music/models.py [hunk 1]
```diff
@@ -54,10 +56,17 @@ def add(cls, **kwargs):
                 existing_music.timer_end = kwargs['timer_end']
             existing_music.date = datetime.now()
             existing_music.save()
+            if not existing_music.room.current_music == existing_music:
```
Comment (Amoki) [inline music/models.py]: existing_music.room.current_music != existing_music

File: music/models.py [hunk 2]
```diff
@@ -54,10 +56,17 @@ def add(cls, **kwargs):
...
+                if PlaylistTrack.objects.filter(track__pk=existing_music.pk):
```
Comment (Amoki) [inline music/models.py]: PlaylistTrack.objects.filter(track=existing_music)

File: music/models.py [hunk 3]
```diff
@@ -54,10 +56,17 @@ def add(cls, **kwargs):
...
+                    PlaylistTrack.objects.get(track__pk=existing_music.pk).top()
```
Comment (Amoki) [inline music/models.py]: idem

File: website/views/remote/remote.py [hunk 4]
```diff
@@ -46,6 +46,13 @@ def next_music(request, room):
     return JSONResponse(remote_template_rendered)
 
 
+@api_view(['POST'])
+@room_required
+def change_ordering(request, room):
+    room.order_playlist(id=request.data.get('music_id'), action=request.data.get('action'))
+    return JSONResponse("ok!")
```
Comment (Amoki) [inline website/views/remote/remote.py]: Response(status=status.HTTP_204_NO_CONTENT)

File: player/models.py [hunk 5]
```diff
@@ -222,6 +228,20 @@ def toggle_shuffle(self, to_active):
             self.save()
             self.send_update_message()
 
+    def order_playlist(self, id, action):
+        print("action : " + action + " on : " + id)
+        if action == "top":
```
Comment (Amoki) [inline player/models.py]: if action no in ['top', 'up', 'down', 'bot]:
    return
playlist = PlaylistTrack.objects.get(room=self, track__music_id=id)
getattr(playlist, action)()

File: music/models.py [hunk 6]
```diff
@@ -54,10 +56,17 @@ def add(cls, **kwargs):
                 existing_music.timer_end = kwargs['timer_end']
             existing_music.date = datetime.now()
             existing_music.save()
+            if existing_music.room.current_music != existing_music:
+                if PlaylistTrack.objects.filter(track=existing_music):
```
Comment (Amoki) [inline music/models.py]: use try catch

File: player/models.py [hunk 7]
```diff
@@ -180,12 +181,12 @@ def get_current_time_past_percent(self):
...
     def get_musics_remaining(self):
         if self.current_music:
-            return self.music_set.filter(date__gt=self.current_music.date).order_by('date')
+            return self.tracks.all().order_by('playlisttrack__order')
         return []
 
     def get_count_remaining(self):
         if self.current_music:
-            return self.music_set.filter(date__gte=self.current_music.date).count()
+            return self.tracks.all().count()
```
Comment (Amoki) [inline player/models.py]: return self.tracks.count()

**Comment by Eirika** [comment]
Rdy to merge, backend is functional (unless you want to move the model PlaylistTrack outside the Music app but personnaly I don't know how without breacking everything each time I try to :-1:  ), front-end will be refactored by KnockoutJS

**Pull request merged**